src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -17,123 +21,170 @@ class MSMonitor(MSCore):
     MSMonitor class provide whole logic behind
     the transferor monitoring module.
     """
+    def __init__(self, msConfig, logger=None):
+        super(MSMonitor, self).__init__(msConfig, logger)
+        # update interval is used to check records in CouchDB and update them
+        # after this interval, default 6h
+        self.updateInterval = self.msConfig.get('updateInterval', 6*60*60)
+        self.campaigns = {}
+
+    def updateCaches(self):
+        """
+        Fetch some data required for the monitoring logic, e.g.:
+         * all campaign configuration
+        :return: True if all of them succeeded, else False
+        """
+        campaigns = self.reqmgrAux.getCampaignConfig("ALL_DOCS")
+        if not campaigns:
+            self.logger.warning("Failed to fetch the campaign configurations")
+        else:
+            # let's make campaign look-up easier and more efficient
+            self.campaigns = {}
+            for camp in campaigns:
+                if 'campaignName' not in camp:
+                    self.logger.warning(
+                            'No campaignName attribute in campaign dict: %s',
+                            json.dumps(camp))
+                    continue
+                self.campaigns[camp['campaignName']] = camp
+            return True
+        return False
+
     def execute(self, reqStatus):
         """
-        Executes the MS monitoring logic
+        Executes the MS monitoring logic, see
+        https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Monitor
+
         :param reqStatus: request statue to process
         :return:
         """
+        if not self.updateCaches():
+            # then wait until the next cycle
+            msg = "Failed to fetch data from one of the data sources. Retrying again in the next cycle"
+            self.logger.error(msg)
+            return
+
         try:
             # get requests from ReqMgr2 data-service for given statue
             # here with detail=False we get back list of records
             requests = self.reqmgr2.getRequestByStatus([reqStatus], detail=False)
             self.logger.debug('+++ monit found %s requests in %s state',
                               len(requests), reqStatus)
 
-            # FIXME: completion threshold should come either from unified
-            # or campaign configuration, e.g.
-            # thr = self.unifiedConfig().get('completedThreshold', 100)
-            thr = 100
-
-            requestStatuses = {}  # keep track of request statuses
+            # get list of all transfer records
+            transferRecords = [d for d in self.getTransferInfo('ALL_DOCS')]
+            # keep track of completed workflows
+            completedWorkflows = []
+            # keep track of request and their new statuses
+            reqStatus = {}
+            # main logic
             for reqName in requests:
```

Comment:
It might give us a better performance if we iterate over the `transferRecords` and check if the workflow name is in `requests`. If we do so, we need to have a mechanism which reports workflows in staging status (in `requests`) but without a correspondent transfer document though.